### ISSUE TITLE
handle DOM scripts

### ISSUE THREAD

**Issue by paulpruteanu** [cypress-io/cypress#828]
Is there a chance to manipulate the scripts that are requested on a page?
The reason I'm asking is that, from my observations, my tests are considerably slow when scripts are requesting 3rd party images: cy.contains/ cy.get will wait for all these resources to be loaded, which are irrelevant to my test cases. A manual remove of those script files significantly increases the performance, and I was thinking maybe I can prohibit the load from within the test.

Thank you!

**Comment by jennifer-shehane** [comment]
You could conditionally load scripts when running tests by checking that Cypress is running within your application code:
https://docs.cypress.io/faq/questions/using-cypress-faq.html#Is-there-any-way-to-detect-if-my-app-is-running-under-Cypress

**Comment by paulpruteanu** [comment]
Thanks, but I'd rather keep my app testing framework agnostic. I think I found a way to do it meanwhile, using cy.document().then((doc) => { /* find script and remove it */ }).

**Comment by brian-mann** [comment]
I don't think I understand the situation.

`cy.contains` and `cy.get` are not blocked by anything - especially external HTTP requests that come from images.

I think you're describing `cy.visit` - which does wait until the `load` event occurs. The `load` event is blocked by all `script`, `link`, `img`, etc.

But even with that said - you couldn't get any benefit by removing those *after* the visit has completed because by then the `load` event has already occurred.

Maybe I'm missing something or there is an edge case I don't understand in regards to `cy.contains` and `cy.get`.

One use case and possible solution for things like Google Analytics is to prevent it from ever loading to begin with. Yes, the script will still make an HTTP request and will be downloaded and parsed but it will not actually send data, which may speed things up.

Here's an example: https://docs.cypress.io/api/events/catalog-of-events.html#Window-Before-Load

**Comment by paulpruteanu** [comment]
Correct, it's cy.visit that's before each test, and that's how my cy.contains() gets delayed by consequence. Indeed, cy.document() comes too late in the picture, I'd look for options to use beforeLoad.

Otherwise, blocking any request (including such a javascript resource, or img, etc. would simplify the approach, which is already in your roadmap.

**Comment by brian-mann** [comment]
You could use a service worker inside of your Cypress test code to accomplish this now if you wanted to.

**Comment by demetriusnunes** [comment]
@brian-mann I'm having intermittent failing tests because of the same issue. Would you elaborate on the suggestion above to resolve it?

**Comment by brian-mann** [comment]
@demetriusnunes what same issue? You'll need to explain your use case / provide a reproducible example, etc.

Something like:

I'm seeing failures due to X.

I've tried Y.

Doing Z would solve my issue.

**Comment by demetriusnunes** [comment]
Ok. I run around 50 E2E tests every hour on my CI using Cypress pointing to my production environment (which loads tons of 3rd party resources via Google Tag Manager, like ad networks, social networks scripts, etc). About 20% of the time, at least 1 or 2 tests fail because the page load event on `cy.visit` didn't fire when some of these 3rd party requests became forever pending. Usually it happens to test suites that have `cy.visit` placed on the `beforeEach` hook. Don't know if this is a coincidence or not, but the failing tests are usually the same 2 or 3. So I wanted to reduce the frequency of these failures by blocking these 3rd party requests.

**Comment by zynchro** [comment]
@demetriusnunes stale issue - but just wondering if you found a nice solution to this. I am facing the same exact scenario with GTM holding up my `cy.visit()`

**Comment by jennifer-shehane** [comment]
@demetriusnunes @zynchro You may want to have the GTM url passed to `blacklistHosts` to block those requests: https://on.cypress.io/configuration#blacklistHosts

**Comment by zynchro** [comment]
@jennifer-shehane thanks for the tip! I'll check it out, seems like a nicer alternative.

For the time being, `before:load` (as mentioned above) worked:
```JS
  const stub = cy.stub();
  cy.on("window:before:load", win => {
    Object.defineProperty(win, "snaptr", {
      configurable: false,
      writeable: false,
      get: () => stub
    });
  });
```

Seemed like a specific tracker which was being loaded via GTM, was the culprit (`snaptr` above)
Similarly, "ga" or "gtm" could be blocked as well.

<details>
<summary>Example with more context</summary>

```JS
function preventTracking() {
  const stub = cy.stub();
  cy.on("window:before:load", win => {
    Object.defineProperty(win, "snaptr", {
      configurable: false,
      writeable: false,
      get: () => stub
    });
    Object.defineProperty(win, "ga", {
      configurable: false,
      writeable: false,
      get: () => stub
    });
  });
}

describe("Site Navigation", () => {
  function basics() {
    context("Basics", () => {
      it("Loads the homepage", () => {
        preventTracking();
        cy.visit("https://www.example.com/");
        cy.title().should("include", "noon");
        ...
```
</details>

**Comment by jennifer-shehane** [comment]
@wrotson No, not currently. See https://github.com/cypress-io/cypress/issues/3820
